### PR TITLE
Add VRRP Layer

### PR DESCRIPTION
**Pull request by wangchong2023**
1. Add VRRP Protocol; 2.Optimize pseudo header checksum.

### CODE REVIEW

**Comment by wangchong2023** [comment]
@seladb 
Hello, I'm angelyouyou(another account).
I don't know what to do with these errors, please help me.

image


image

**Comment by wangchong2023** [comment]
@seladb 
Hello, I‘m angelyouyou(another account).
CI always prompts a document construction error during pre-build, but there is no information in the error details.What should I do next?

image

**Comment by seladb** [comment]
> @seladb Hello, I‘m angelyouyou(another account). CI always prompts a document construction error during pre-build, but there is no information in the error details.What should I do next?
> 
> image
> 
> image

Hi @wangchong2023 ! Please notice that the documentation is built in Cirrus CI, so you can click this link to see more details:

image

Once you click you can see the full log: https://cirrus-ci.com/task/6097118847303680

Notice this error here:
```text
Generating docs for compound pcpp:/tmp/cirrus-ci-build/Packet++/header/VrrpLayer.h:188: error: argument 'version' of command @param is not found in the argument list of pcpp::VrrpLayer::VrrpLayer(ProtocolType subProtocol) (warning treated as error, aborting now)
```

File: Common++/header/Logger.h [hunk 1]
```diff
@@ -59,6 +59,7 @@ namespace pcpp
 		PacketLogModuleGtpLayer, ///< GtpLayer module (Packet++)
 		PacketLogModuleBgpLayer, ///< GtpLayer module (Packet++)
 		PacketLogModuleSSHLayer, ///< SSHLayer module (Packet++)
+        PacketLogModuleVrrpLayer, ///< Vrrp Record module (Packet++)
```
Comment (seladb) [inline Common++/header/Logger.h:62]: Please switch from spaces to tabs `\t` everywhere (not just in this file). The current convention is to use tabs

File: README.md [hunk 2]
```diff
@@ -226,38 +226,55 @@ PcapPlusPlus currently supports parsing, editing and creation of packets of the
...
-27. UDP
+24. GTP (v1)
+
+25. IPSec AH & ESP - parsing only (no editing capabilities)
+
+26. TCP
+
+27. TPKT
+
+28. UDP
```
Comment (seladb) [inline README.md:241]: Why did you add these extra spaces?

Comment (wangchong2023) [reply README.md:241]: Maybe Clion added these. I do not know. I will check about it.

File: Common++/header/Logger.h (hunk 1 above)
Comment (wangchong2023) [reply Common++/header/Logger.h:62]: OK.I will fix it.

**Comment by wangchong2023** [comment]
@seladb 
I have modified it according to the opinions you just mentioned, please review the code again, thank you.
I do n’t know how to repair CodeCov's error. I do n’t know how to view the specific decline point? If you need to repair, please let me know.

If there is no problem, please merge the PR.Thanks.

File: Packet++/header/VrrpLayer.h [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+		 * A method that gets VRRP authentication type description
+		* @return The VRRP authentication type description
+		*/
+		std::string getAuthTypeDesc() const;
+
+		/**
+		* A method that gets VRRP maximum advertisement interval
+		* @return The maximum advertisement interval in this message
+		*/
+		uint8_t getAdvInt() const;
```
Comment (seladb) [inline Packet++/header/VrrpLayer.h:544]: If I understand correctly in VRRP v3 the AdvInt size is bigger than `unit8_t` (8 bit), it's actually 12-bit: https://datatracker.ietf.org/doc/html/rfc5798

I think that instead of creating a virtual `getAdvInt()` method in `VrrpLayer` and then override it in `VrrpV2Layer` and `VrrpV3Layer` you can just implement 2 separate methods in both classes, each one with a slightly different signature (in v2 it'll return `uint8_t` and in v3 it'll return `uint16_t`)

File: Packet++/header/VrrpLayer.h [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+	private:
+		/**
+		* A method that add IP Addresses to data offset
+		* @param ipAddresses [in] IP Address vector
+		* @param offset [in] Data offset
+		* @return ture if add successfully, false otherwise.
+		*/
+		bool addIPAddressesAt(const std::vector<IPAddress> &ipAddresses, int offset);
+
+		IPAddress::AddressType m_addressType;
```
Comment (seladb) [inline Packet++/header/VrrpLayer.h:177]: very nit: the variable name should be `m_AddressType`

File: Packet++/header/VrrpLayer.h [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+
+	public:
+		VrrpLayer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet, ProtocolType vrrpVer,
+				  IPAddress::AddressType addressType)
+				: Layer(data, dataLen, prevLayer, packet) {
+			m_Protocol = vrrpVer;
+			m_addressType = addressType;
+		}
```
Comment (seladb) [inline Packet++/header/VrrpLayer.h:185]: I think we should make this constructor protected instead of public so only the child classes can call it but not an outside user

File: Packet++/header/VrrpLayer.h [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+
+		/**
+		 * A constructor that allocates a new VRRP
+		 * @param[in] subProtocol The VRRP protocol
+		 */
+		explicit VrrpLayer(ProtocolType subProtocol);
```
Comment (seladb) [inline Packet++/header/VrrpLayer.h:191]: Same here: this constructor should also be protected, that's how we ensure no one can create an instance of `VrrpLayer`

File: Packet++/header/VrrpLayer.h [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+		* A method that get IP Address type
+		* @return The VRRP IP Address type
+		*/
+		IPAddress::AddressType getAddressType() const;
+
+		/**
+		* A method that set IP Address type
+		* @param addressType [in] IP Address type
+		*/
+		void setAddressType(IPAddress::AddressType addressType);
```
Comment (seladb) [inline Packet++/header/VrrpLayer.h:205]: What's the point in exposing a public method to set the address type? 🤔

File: Packet++/src/VrrpLayer.cpp [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+
+		getPacketPtr()->ipAddrCount = 0;
+		getPacketPtr()->version = getVersionFromData();
+
+		m_addressType = IPAddress::IPv4AddressType;
+	}
+
+	bool VrrpLayer::isDataValid(size_t dataLen) {
+		return (dataLen > VRRP_PACKET_FIX_LEN);
+	}
```
Comment (seladb) [inline Packet++/src/VrrpLayer.cpp:31]: 1. I think this method should accept a pointer to the data and length: `VrrpLayer::isDataValid(uint8_t* data, size_t dataLen)` and then check: `data && dataLen > ...`
2. The minimum data length to parse a packet is `sizeof(vrrp_packet)`, no?

File: Packet++/src/VrrpLayer.cpp [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+		auto *vrrpPacketCopy = (vrrp_packet *) new uint8_t[getDataLen()];
+		memcpy(vrrpPacketCopy, getData(), getDataLen());
+
+		ScalarBuffer<uint16_t> buffer = {};
+		buffer.buffer = (uint16_t *) vrrpPacketCopy;
+		buffer.len = getHeaderLen();
+
+		vrrpPacketCopy->checksum = 0;
+		uint16_t checksum = computeChecksum(&buffer, 1);
+		delete[] vrrpPacketCopy;
```
Comment (seladb) [inline Packet++/src/VrrpLayer.cpp:477]: coping the data is quite expensive. Instead we can save the existing checksum value in a variable, set the checksum to 0, calculate the checksum and then set the value back. This is what we do in `TcpLayer`:
https://github.com/seladb/PcapPlusPlus/blob/3dca454eaf8f8440c3e302df262dc5da637b6c5f/Packet%2B%2B/src/TcpLayer.cpp#L246

File: Packet++/src/VrrpLayer.cpp [hunk 10]
```diff
@@ -0,0 +1,533 @@
...
+
+	uint16_t VrrpV2Layer::fillChecksum() {
+		auto *vrrpPacket = (vrrp_packet *) getData();
+		ScalarBuffer<uint16_t> buffer = {};
+		buffer.buffer = (uint16_t *) getData();
+		buffer.len = getHeaderLen();
+
+		vrrpPacket->checksum = 0;
+		uint16_t checksum = computeChecksum(&buffer, 1);
+		vrrpPacket->checksum = htobe16(checksum);
```
Comment (seladb) [inline Packet++/src/VrrpLayer.cpp:458]: 1. I'd advise renaming this method to `calculateAndSetChecksum()`
2. Instead of duplicating the code between this method and `calculateChecksum()` we can just call it:
   ```cpp
   auto *vrrpPacket = (vrrp_packet *) getData();
   vrrpPacket->checksum = htobe16(calculateChecksum());
   ```

File: Packet++/src/VrrpLayer.cpp [hunk 11]
```diff
@@ -0,0 +1,533 @@
...
+			checksum = computePseudoHdrChecksum((uint8_t *) vrrpPacketCopy, getDataLen(), IPAddress::IPv4AddressType,
+												PACKETPP_IPPROTO_VRRP, srcIPAddr, dstIPAddr);
+		} else {
+			checksum = computePseudoHdrChecksum((uint8_t *) vrrpPacketCopy, getDataLen(), IPAddress::IPv6AddressType,
+												PACKETPP_IPPROTO_VRRP, srcIPAddr, dstIPAddr);
+		}
+
+		delete[]vrrpPacketCopy;
+
+		return checksum;
```
Comment (seladb) [inline Packet++/src/VrrpLayer.cpp:531]: Please see my previous comments in `VrrpV2Layer` regarding these 2 methods

File: Packet++/src/VrrpLayer.cpp [hunk 12]
```diff
@@ -0,0 +1,533 @@
...
+		Virtual Rtr ID: 1
+		Priority: 100 (Default priority for a backup VRRP router)
+		Addr Count: 1
+		Auth Type: No Authentication (0)
+		Adver Int: 1
+		Checksum: 0xba52 [correct]
+		IP Address:
+			192.168.0.1
+	* */
+	std::string VrrpLayer::toString() const {
```
Comment (seladb) [inline Packet++/src/VrrpLayer.cpp:142]: `toString()` should consist of one line that describes the most important data of a layer, a little bit like Wireshark does:
image

We should probably remove most of the data from here and keep just the version and maybe some more super important data, but definitely not all of it

File: Packet++/src/VrrpLayer.cpp [hunk 13]
```diff
@@ -0,0 +1,533 @@
...
+			priorityName = "(Default priority for a backup VRRP router)";
+		} else if (getPriority() == VRRP_PRIO_STOP) {
+			priorityName = "(Current Master has stopped participating in VRRP)";
+		} else if (getPriority() == VRRP_PRIO_OWNER) {
+			priorityName = "(This VRRP router owns the virtual router's IP address(es))";
+		} else {
+		}
+
+		return priorityName;
+	}
```
Comment (seladb) [inline Packet++/src/VrrpLayer.cpp:112]: if we make `toString()` shorter this method is probably not needed

File: Packet++/src/VrrpLayer.cpp [hunk 14]
```diff
@@ -0,0 +1,533 @@
...
+			priorityName = "Simple Text Authentication [RFC 2338] / Reserved [RFC 3768]";
+		} else if (authType == VRRP_AUTH_AH) {
+			priorityName = "IP Authentication Header [RFC 2338] / Reserved [RFC 3768]";
+		} else if (authType == VRRP_AUTH_MD5) {
+			priorityName = "Cisco VRRP MD5 authentication";
+		} else {
+		}
+
+		return priorityName;
+	}
```
Comment (seladb) [inline Packet++/src/VrrpLayer.cpp:128]: if we make `toString()` shorter this method is probably not needed

File: Packet++/src/VrrpLayer.cpp [hunk 15]
```diff
@@ -0,0 +1,533 @@
...
+	std::string VrrpLayer::getTypeName() const {
+		std::string typeName;
+		if (getPacketPtr()->type == VrrpType_Advertisement) {
+			typeName = "Advertisement";
+		} else {
+			typeName = "Unknown";
+		}
+
+		return typeName;
+	}
```
Comment (seladb) [inline Packet++/src/VrrpLayer.cpp:98]: if we make `toString()` shorter this method is probably not needed

**Review by seladb — COMMENTED** [review]
@wangchong2023 I didn't finish reviewing the entire PR, but I feel there are enough comments and it's better if you fix them first and then I'll continue reviewing

File: Tests/Packet++Test/PacketExamples/TCP-IPv4-Checksum-Bad.dat
[the remark this answers is not in the document]
Comment (wangchong2023) [reply Tests/Packet++Test/PacketExamples/TCP-IPv4-Checksum-Bad.dat:1]: OK.I will add it.

File: Tests/Packet++Test/main.cpp [hunk 16]
```diff
@@ -306,5 +306,8 @@ int main(int argc, char* argv[])
 	PTF_RUN_TEST(WakeOnLanCreationTests, "wol");
 	PTF_RUN_TEST(WakeOnLanEditTests, "wol");
 
+    PTF_RUN_TEST(VrrpParsingTest, "vrrp");
```
[the remark this answers is not in the document]
Comment (wangchong2023) [reply Tests/Packet++Test/main.cpp:309]: OK. In my local Clion, I format them all. I will check it agin. Is there a tool for format on line in this project?
@seladb

File: Packet++/src/VrrpLayer.cpp [hunk 17]
```diff
@@ -0,0 +1,533 @@
...
+										 IPAddress::AddressType addressType) {
+		auto *vrrpPacketCommon = (vrrp_packet *) data;
+		uint8_t version = vrrpPacketCommon->version;
+		switch (version) {
+			case Vrrp_Version_2:
+				return new VrrpV2Layer(data, dataLen, prevLayer, packet);
+			case Vrrp_Version_3:
+				return new VrrpV3Layer(data, dataLen, prevLayer, packet, addressType);
+			default: {
+				return nullptr;
```
[the remark this answers is not in the document]
Comment (wangchong2023) [reply Packet++/src/VrrpLayer.cpp:43]: OK. I will fix it.

File: Packet++/header/ProtocolType.h [hunk 18]
```diff
@@ -302,6 +302,20 @@ namespace pcpp
...
+	/**
+	 * VRRP version 2 protocol
+	 */
+	const ProtocolType VRRPv2 = 0x1000000000000;
+
+	/**
+	 * VRRP version 3 protocol
+	 */
+	const ProtocolType VRRPv3 = 0x2000000000000;
 	/**
```
[the remark this answers is not in the document]
Comment (wangchong2023) [reply Packet++/header/ProtocolType.h:319]: OK.

File: Packet++/header/VrrpLayer.h [hunk 19]
```diff
@@ -0,0 +1,560 @@
...
+#define VRRP_AUTH_NONE      0     /* no authentication */
+#define VRRP_AUTH_SIMPLE    1     /* Simple Text Authentication [RFC 2338] */
+#define VRRP_AUTH_AH        2     /* IP Authentication Header [RFC 2338]  */
+#define VRRP_AUTH_MD5       3     /* Cisco VRRP MD5 authentication  */
+
+#define VRRP_PACKET_FIX_LEN 8
+#define VRRP_PACKET_MAX_IP_ADDRESS_NUM 255
+
+#define VRRP_PACKET_SET_AUTH_ADV_INT(authTypeAdvInt, authAdvInt) ((authTypeAdvInt) = *((uint16_t*)&(authAdvInt)))
+
```
[the remark this answers is not in the document]
Comment (wangchong2023) [reply Packet++/header/VrrpLayer.h:36]: OK. I will remove the useless define.

File: Packet++/header/VrrpLayer.h [hunk 20]
```diff
@@ -0,0 +1,560 @@
...
+		 0                   1                   2                   3
+		 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+		+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+		|                    IPv4 Fields or IPv6 Fields                 |
+	   ...                                                             ...
+		|                                                               |
+		+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+		|Version| Type  | Virtual Rtr ID|   Priority    |Count IPvX Addr|
+		+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+		|(varLenFlag) |     Max Adver Int     |          Checksum             |
```
[the remark this answers is not in the document]
Comment (wangchong2023) [reply Packet++/header/VrrpLayer.h:80]: OK.I will fix it.

File: Packet++/header/VrrpLayer.h (hunk 4 above)
Comment (wangchong2023) [reply Packet++/header/VrrpLayer.h:177]: OK. I will fix it.

File: Packet++/header/VrrpLayer.h (hunk 5 above)
Comment (wangchong2023) [reply Packet++/header/VrrpLayer.h:185]: OK.I will fix it.

File: Packet++/src/VrrpLayer.cpp (hunk 13 above)
Comment (wangchong2023) [reply Packet++/src/VrrpLayer.cpp:112]: The macro define is used here. I add this function for short function.

File: Packet++/src/VrrpLayer.cpp (hunk 15 above)
Comment (wangchong2023) [reply Packet++/src/VrrpLayer.cpp:98]: The macro define is used here. I add this function for short function for extension.

File: Packet++/header/VrrpLayer.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (wangchong2023) [reply Packet++/header/VrrpLayer.h:36]: Yes, these macros are used in VrrpLayer.cpp and VrrpTest.cpp. I add these macros to avoid magic data.

File: Packet++/header/VrrpLayer.h (hunk 6 above)
Comment (wangchong2023) [reply Packet++/header/VrrpLayer.h:191]: OK.I will fix it.

File: Packet++/src/VrrpLayer.cpp (hunk 8 above)
Comment (wangchong2023) [reply Packet++/src/VrrpLayer.cpp:31]: Yes, you are right. I have add function 'getVersionFromData' to change the parse action.

File: Packet++/src/VrrpLayer.cpp (hunk 10 above)
Comment (wangchong2023) [reply Packet++/src/VrrpLayer.cpp:458]: OK. I will fix it. After that please review again.

File: Packet++/src/VrrpLayer.cpp (hunk 9 above)
Comment (wangchong2023) [reply Packet++/src/VrrpLayer.cpp:477]: OK. I will fix it.

File: Packet++/src/VrrpLayer.cpp (hunk 11 above)
Comment (wangchong2023) [reply Packet++/src/VrrpLayer.cpp:531]: OK. I will fix it.

File: Packet++/header/VrrpLayer.h (hunk 3 above)
Comment (wangchong2023) [reply Packet++/header/VrrpLayer.h:544]: OK. I will fix it.

File: Tests/Packet++Test/PacketExamples/TCP-IPv4-Checksum-Bad.dat
[the remark this answers is not in the document]
Comment (seladb) [reply Tests/Packet++Test/PacketExamples/TCP-IPv4-Checksum-Bad.dat:1]: @wangchong2023 this file was not added...

File: Tests/Packet++Test/main.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (seladb) [reply Tests/Packet++Test/main.cpp:309]: Unfortunately there is no formatting tool for this project yet

File: Packet++/src/VrrpLayer.cpp [hunk 21]
```diff
@@ -0,0 +1,517 @@
...
+		m_Data = new uint8_t[m_DataLen];
+		memset(m_Data, 0, m_DataLen);
+		m_Protocol = subProtocol;
+		m_AddressType = IPAddress::IPv4AddressType;
+	}
+
+	ProtocolType VrrpLayer::getVersionFromData(uint8_t *data, size_t dataLen) {
+		if (dataLen <= VRRP_PACKET_FIX_LEN) {
+			return UnknownProtocol;
+		}
```
Comment (seladb) [inline Packet++/src/VrrpLayer.cpp:28]: This check should be:
```cpp
if (!data || dataLen < sizeof(vrrp_packet))
{
	return UnknownProtocol;
}
```

File: Packet++/header/VrrpLayer.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/header/VrrpLayer.h:35]: I'm not an expert in the VRRP protocol, but maybe `toString()` can be something like this:
```text
// for VRRP v2:
VRRP v2 Layer, virtual router ID: 5, IP address count: 3

// for VRRP v3:
VRRP v3 Layer, virtual router ID: 5, IP address count: 3
```

File: Packet++/header/VrrpLayer.h (hunk 3 above)
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: I'm not sure this method must be on the parent, we can define it separately on `VrrpV2Layer` and `VrrpV3Layer` with a slightly different signature (in v2 it'll return `uint8_t` and in v3 it'll return `uint16_t`)

File: Packet++/src/VrrpLayer.cpp (hunk 15 above)
Comment (seladb) [reply Packet++/src/VrrpLayer.cpp:93]: Please see this comment: https://github.com/seladb/PcapPlusPlus/pull/1118#discussion_r1165076754

File: Packet++/src/VrrpLayer.cpp (hunk 12 above)
Comment (seladb) [reply Packet++/src/VrrpLayer.cpp:137]: Please see this comment: https://github.com/seladb/PcapPlusPlus/pull/1118#discussion_r1165076754

File: Packet++/header/VrrpLayer.h [hunk 22]
```diff
@@ -0,0 +1,545 @@
...
+
+		/**
+		 * Get the virtual IP address that comes next to a given virtual IP address. If "ipAddress" is NULL then NULL will be returned.
+		 * If "ipAddress" is the last virtual IP address or if it is out of layer bounds NULL will be returned also. Notice the return value is a
+		 * pointer to the real data casted to IPAddress type (as opposed to a copy of the option data). So changes in the return
+		 * value will affect the packet data
+		 * @param[in] ipAddress The virtual IP address to start searching from
+		 * @return The next virtual IP address or NULL if "ipAddress" is NULL, last or out of layer bounds
+		 */
+		uint8_t *getNextIPAddress(uint8_t *ipAddress) const;
```
Comment (seladb) [inline Packet++/header/VrrpLayer.h:347]: I'm not sure these methods are really needed. And even if they are needed we should make them private and rename them to `getFirstIPAddressPtr()` and `getNextIPAddressPtr()`.

Essentially all they do is check `getIPAddressLen()` which always returns the same value for a specific packet (so we only need to call it once before starting iterating over the addresses), and then they just move the pointer.

I don't think we need specific methods for that.

File: Packet++/header/VrrpLayer.h [hunk 23]
```diff
@@ -0,0 +1,545 @@
...
+		 * @param[in] data A data pointer
+		 * @param[in] ipAddress IP address
+		 */
+		void copyIPAddressToData(uint8_t *data, const IPAddress &ipAddress) const;
+
+		/**
+		 * Get IP address from data
+		 * @return true if get successfully, otherwise return false.
+		 */
+		bool getIPAddressFromData(uint8_t *data, IPAddress &ipAddress) const;
```
Comment (seladb) [inline Packet++/header/VrrpLayer.h:391]: This method should be private

File: Packet++/src/VrrpLayer.cpp [hunk 24]
```diff
@@ -0,0 +1,517 @@
...
+			ipAddress = ipv4Address;
+		} else if (getAddressType() == IPAddress::IPv6AddressType) {
+			IPv6Address ipv6Address(data);
+			ipAddress = ipv6Address;
+		} else {
+			return false;
+		}
+
+		return true;
+	}
```
Comment (seladb) [inline Packet++/src/VrrpLayer.cpp:389]: A nicer implementation for this method:
```cpp
IPAddress VrrpLayer::getIPAddressFromData(uint8_t *data) const {
	if (getAddressType() == IPAddress::IPv4AddressType) {
		return IPv4Address(*((uint32_t *) data));
	} else {
		return IPv6Address(data);
}
```

**Review by seladb — COMMENTED** [review]
This is another round of review, but I didn't finish reviewing everything. Please fix these comments and I'll continue the review

**Comment by seladb** [comment]
@wangchong2023 will you have time to continue working on this PR?
All the comments that are still open (not resolved) weren't addressed yet

File: Tests/Packet++Test/main.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (seladb) [reply Tests/Packet++Test/main.cpp]: Fixed here: https://github.com/seladb/PcapPlusPlus/pull/1118/commits/c67f8a342fa10a277c204309540c2825a53d5da3

File: Packet++/src/IPv4Layer.cpp [hunk 25]
```diff
@@ -246,6 +247,7 @@ void IPv4Layer::parseNextLayer()
 
 	ProtocolType greVer = UnknownProtocol;
 	ProtocolType igmpVer = UnknownProtocol;
+	ProtocolType vrrpVer = UnknownProtocol;
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/src/IPv4Layer.cpp]: Fixed here: https://github.com/seladb/PcapPlusPlus/pull/1118/commits/c67f8a342fa10a277c204309540c2825a53d5da3

File: Packet++/header/VrrpLayer.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed here: https://github.com/seladb/PcapPlusPlus/pull/1118/commits/ddda563d5ded2fb531129b1ce85db68b578f39d1

File: Packet++/src/VrrpLayer.cpp (hunk 12 above)
Comment (seladb) [reply Packet++/src/VrrpLayer.cpp]: fixed here: https://github.com/seladb/PcapPlusPlus/pull/1118/commits/ddda563d5ded2fb531129b1ce85db68b578f39d1

File: Packet++/src/VrrpLayer.cpp (hunk 15 above)
Comment (seladb) [reply Packet++/src/VrrpLayer.cpp]: Removed here: https://github.com/seladb/PcapPlusPlus/pull/1118/commits/ddda563d5ded2fb531129b1ce85db68b578f39d1

File: Packet++/header/VrrpLayer.h [hunk 26]
```diff
@@ -0,0 +1,545 @@
...
+		* A method that gets VRRP checksum
+		* @return The checksum in this message
+		*/
+		uint16_t getChecksum() const;
+
+		/**
+		 * A method that gets count of VRRP virtual IP addresses
+		 * @return The count of virtual IP addresses in this message
+		 */
+		uint16_t getIPAddressesCount() const;
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/6bcaa96a4cf7117ccb736e5224f0ea178d6bf1a3

File: Packet++/src/VrrpLayer.cpp [hunk 27]
```diff
@@ -0,0 +1,517 @@
...
+
+	uint8_t VrrpLayer::getVersion() const {
+		vrrp_packet *packet = getPacketPtr();
+		return packet->version;
+	}
+
+	VrrpType VrrpLayer::getType() const {
+		vrrp_packet *packet = getPacketPtr();
+		return (VrrpType) (packet->type);
+	}
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/src/VrrpLayer.cpp]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/6bcaa96a4cf7117ccb736e5224f0ea178d6bf1a3

File: Packet++/src/VrrpLayer.cpp [hunk 28]
```diff
@@ -0,0 +1,517 @@
...
+
+	uint8_t VrrpLayer::getPriority() const {
+		vrrp_packet *packet = getPacketPtr();
+		return packet->priority;
+	}
+
+	uint16_t VrrpLayer::getChecksum() const {
+		auto *packet = (vrrp_packet *) getData();
+		return packet->checksum;
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/src/VrrpLayer.cpp]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/6bcaa96a4cf7117ccb736e5224f0ea178d6bf1a3

File: Packet++/header/VrrpLayer.h [hunk 29]
```diff
@@ -0,0 +1,545 @@
...
+		 * VRRP version
+		 */
+		uint8_t getVersionFromData() const;
+
+		/**
+		* A method that gets VRRP version
+		* @return The version in this message
+		*/
+		uint8_t getVersion() const;
+
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/6bcaa96a4cf7117ccb736e5224f0ea178d6bf1a3

File: Packet++/header/VrrpLayer.h [hunk 30]
```diff
@@ -0,0 +1,545 @@
...
+		 * Get a pointer to the raw VRRPv2/VRRPv3 header. Notice this points directly to the data, so every change will change the actual packet data
+		 * @return A pointer to the @ref vrrp_packet
+		 */
+		vrrp_packet *getPacketPtr() const { return (vrrp_packet *) m_Data; }
+
+		/**
+		 * A method that gets length of virtual address
+		 * @return Virtual IP address length
+		 */
+		uint8_t getIPAddressLen() const;
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/6bcaa96a4cf7117ccb736e5224f0ea178d6bf1a3

File: Packet++/src/VrrpLayer.cpp [hunk 31]
```diff
@@ -0,0 +1,517 @@
...
+
+	std::string VrrpLayer::getTypeName() const {
+		std::string typeName;
+		if (getPacketPtr()->type == VrrpType_Advertisement) {
+			typeName = "Advertisement";
+		} else {
+			typeName = "Unknown";
+		}
+
+		return typeName;
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/src/VrrpLayer.cpp]: Not relevant anymore

File: Packet++/header/VrrpLayer.h [hunk 32]
```diff
@@ -0,0 +1,545 @@
...
+		 * @return IPv4 address of VRRP packet
+		 */
+		uint8_t isIPv4Packet() const;
+
+		/**
+		 * Judge whether the packet is IPv6
+		 * @return IPv6 address of VRRP packet
+		 */
+		uint8_t isIPv6Packet() const;
+
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/6bcaa96a4cf7117ccb736e5224f0ea178d6bf1a3

File: Packet++/src/IPv6Layer.cpp [hunk 33]
```diff
@@ -218,6 +217,8 @@ void IPv6Layer::parseNextLayer()
 		nextHdr = getIPv6Header()->nextHeader;
 	}
 
+	ProtocolType vrrpVer;
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/src/IPv6Layer.cpp]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/6bcaa96a4cf7117ccb736e5224f0ea178d6bf1a3

File: Packet++/src/VrrpLayer.cpp (hunk 21 above)
Comment (seladb) [reply Packet++/src/VrrpLayer.cpp]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/6bcaa96a4cf7117ccb736e5224f0ea178d6bf1a3

File: Packet++/header/VrrpLayer.h (hunk 3 above)
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/71468b7f0c5c05bb3f6c6f9cc96c47ebbb448ef7

File: Packet++/header/VrrpLayer.h [hunk 34]
```diff
@@ -0,0 +1,545 @@
...
+		 * @return True if the checksum is correct
+		 */
+		bool isChecksumCorrect() const;
+
+		/**
+		* Set packet to layer
+		* @param[in] packet The packet to be attached
+		* @return The VRRP authentication type description
+		*/
+		void setPacket(vrrp_packet *packet);
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/71468b7f0c5c05bb3f6c6f9cc96c47ebbb448ef7

File: Packet++/header/VrrpLayer.h [hunk 35]
```diff
@@ -0,0 +1,545 @@
...
+		Vrrp_Version_3 = 3
+	};
+
+	/**
+	 * VRRP message types
+	 */
+	enum VrrpType {
+		/** VRRP advertisement packet */
+		VrrpType_Advertisement = 1
+	};
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/a5fb5e7227817d62480981cef55b88c2a6d22b36

File: Packet++/header/VrrpLayer.h [hunk 36]
```diff
@@ -0,0 +1,545 @@
...
+		 * @return true if get successfully, otherwise return false.
+		 */
+		bool getIPAddressFromData(uint8_t *data, IPAddress &ipAddress) const;
+
+		/**
+		 * Judge whether the IP address is valid
+		 * @param[in] ipAddress IP address
+		 * @return true if ipAddress is valid, otherwise return false.
+		 */
+		bool isIPAddressValid(IPAddress &ipAddress) const;
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/9fbadc5322818d2c02831a7aac01b3577fa14267

File: Packet++/header/VrrpLayer.h (hunk 7 above)
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/9fbadc5322818d2c02831a7aac01b3577fa14267

File: Packet++/src/VrrpLayer.cpp (hunk 24 above)
Comment (seladb) [reply Packet++/src/VrrpLayer.cpp]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/0233d684d8aff42029b747845745d835aa066beb

File: Packet++/header/VrrpLayer.h (hunk 23 above)
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/0233d684d8aff42029b747845745d835aa066beb

File: Packet++/header/VrrpLayer.h (hunk 22 above)
Comment (seladb) [reply Packet++/header/VrrpLayer.h]: I decided to keep these methods and move them to be private. Fix in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/0233d684d8aff42029b747845745d835aa066beb

**Review by seladb — APPROVED** [review]
